jax/_src/nn/functions.py

```diff
@@ -765,3 +768,204 @@ def hard_silu(x: ArrayLike) -> Array:
   return x_arr * hard_sigmoid(x_arr)
 
 hard_swish = hard_silu
+
+def _causal_mask(T, dtype):
+  dtype_max = jnp.finfo(dtype).max
+  large_negative_number = jnp.asarray(-0.5 * dtype_max, dtype=dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+class ScaledDotProductAttentionCausalMask:
+  def __call__(self, x: ArrayLike, mask: ArrayLike | None = None) -> Array:
+    """Mask the input with causal masking"""
+    if mask is not None:
+      warnings.warn(
+          "The provided mask will be ignored since the causal mask is used")
+      del mask
+    T = x.shape[-2]
+    assert T == x.shape[-1], f'T should equal to S: {T} vs {x.shape[-1]}'
+    assert jnp.issubdtype(x.dtype, jnp.floating), x.dtype
+    mask = _causal_mask(T, x.dtype)
+    return x + mask
+
+@dataclass
+class ScaledDotProductAttentionPhiloxDropout:
```

Comment:
What is the sharding behaviour of this dropout? I'm worried that it won't compose well with JAX sharding APIs.

Reply:
To be explicit: consider the data parallel case. Applying this function on a single device should have the same result as applying it sharded over two. This obviously causes issues for the RNG, as per-device, the random tensors are now different shapes. Which keys to use? JAX handles this, but its a problem when using built-in custom-op RNG like in this case.

Reply:
Dropout will be added in the next PRs and we will specify the parallel support for it there. Thanks for bring up this concern.